planning/obstacle_avoidance_planner/src/node.cpp

```diff
@@ -1295,17 +1302,34 @@ void ObstacleAvoidancePlanner::insertZeroVelocityOutsideDrivableArea(
     const bool is_outside = drivable_area_utils::isOutsideDrivableAreaFromRectangleFootprint(
       traj_point, left_bound, right_bound, vehicle_param_, is_considering_footprint_edges_);
 
-    // only insert zero velocity to the first point outside drivable area
+    // only insert zero velocity to the first point outside drivable area (minus stop margin)
     if (is_outside) {
-      traj_points[i].longitudinal_velocity_mps = 0.0;
+      size_t stop_idx = i;
+      const auto & op_target_point = motion_utils::calcLongitudinalOffsetPoint(
+        traj_points, traj_point.pose.position, -1.0 * vehicle_stop_margin_outside_drivable_area_);
+
+      if (op_target_point) {
+        const auto target_point = op_target_point.get();
+        stop_idx = motion_utils::findNearestSegmentIndex(traj_points, target_point);
+      }
+
+      traj_points[stop_idx].longitudinal_velocity_mps = 0.0;
       debug_data_.stop_pose_by_drivable_area = traj_points[i].pose;
+      debug_data_.stop_pose_by_drivable_area_with_margin = traj_points[stop_idx].pose;
 
       // NOTE: traj_points does not have valid z for efficient calculation of trajectory
       if (!planner_data.path.points.empty()) {
-        const size_t path_idx =
+        const size_t path_idx_1 =
           motion_utils::findNearestIndex(planner_data.path.points, traj_points[i].pose.position);
+
         debug_data_.stop_pose_by_drivable_area->position.z =
-          planner_data.path.points.at(path_idx).pose.position.z;
+          planner_data.path.points.at(path_idx_1).pose.position.z;
+
+        const size_t path_idx_2 = motion_utils::findNearestIndex(
+          planner_data.path.points, traj_points[stop_idx].pose.position);
+
+        debug_data_.stop_pose_by_drivable_area_with_margin->position.z =
+          planner_data.path.points.at(path_idx_2).pose.position.z;
```

Comment:
Could you use more descriptive variable names?

Reply:
I returned it back to original code, as Mehmet suggested to have only 1 vistual wall where the vehicle stops